_posts/2018-10-08-http-part5.markdown

```diff
@@ -0,0 +1,305 @@
+            connection.close()
+            break
+        } catch (e: RequestParseException) {
+            logger.warn("Could not parse a request", e)
+            Response(status = Status.BAD_REQUEST)
+        }
+        output.print(response.toRaw())
+        output.flush()
+        if (request?.headers?.connection == "close") { // 8
+            connection.close()
+        }
+    }
+}
+```
+We create a new connection from accepted socket (3), add it to the connection pool (4) and then handle the connection (5). Last two steps are run in `handlerThreadPool` so our server can immediately accept other connections.
+
+The main changes are in `handleConnection` method. Previously, we used `socket.use { }`  block in which we read a request and sent a response. At the end of this block, the connection was closed. We can’t do that now because we want to keep the connection opened. Instead, we should keep reading requests and sending responses while the connection is open.
+
+After a first request from a connection, our server will go through second iteration of the loop and will block on `Request.fromRaw(input)`, (6) which will block on `input.readLine()` of `BufferedReader`, which internally will block on a socket `read()`. This loop ends if either server or client closes the connection (7) or will send `Connection: close` header (8).
+
+#### Detecting connections closed by client
+Detecting closed connections by our server is easy, we can check `socket.isClosed`. The tricky part for me was to check if a connection was closed by the client. When the client closes a connection, it sends `FIN,ACK` segment and we send the `ACK` back.  At first, I thought that we can just check `socket.isClosed` too, but I was wrong - this flag only checks if *we* closed the socket.
+[It seems](https://stackoverflow.com/a/10241044/1189159) that there is no information about `FIN,ACK` event from the client in Socket API.
+The only indication of it is to try to either use `read()` or `readLine()` on `BufferedReader`.
+
+With that in mind, it is hard to do the following logic - block until new data arrives, then parse the request, if there is no data, close a connection. We cannot use `readLine()` before `Request.fromRaw(input)` because it will consume one line of our `BufferedReader`. We could change signature `fromRaw` to accept the first line as `String` and the rest of a stream as `BufferedReader` but that’s just ugly.
+That’s why for now we can keep things as they were. The beginning of our parsing logic protects us from empty requests anyways.
+``` kotlin
+val requestLine = input.readLine() ?: throw EmptyRequestException()
+```
+If `readLine()` returns null, we can just catch it (7) and close the connection. It might not be the best, it may smell like using exceptions for control flow, but it works. 
+We will try to tackle this problem in the future maybe by using some sort of “one line/character caching BufferedReader”.
```

Comment:
We will try to tackle this problem in the future by using some sort of “one line/character caching BufferedReader”. (removed 'maybe' as it was confusing)